File: shared_resources/src/main/java/com/thewizrd/shared_resources/utils/VersionMigrations.java
package com.thewizrd.shared_resources.utils;

import android.content.Context;
import android.content.pm.PackageInfo;
import android.os.Bundle;
import android.util.Log;

import com.thewizrd.shared_resources.SimpleLibrary;
import com.thewizrd.shared_resources.database.LocationsDatabase;
import com.thewizrd.shared_resources.database.WeatherDatabase;
import com.thewizrd.shared_resources.locationdata.LocationData;
import com.thewizrd.shared_resources.preferences.FeatureSettings;
import com.thewizrd.shared_resources.weatherdata.WeatherAPI;
import com.thewizrd.shared_resources.weatherdata.WeatherManager;

class VersionMigrations {
    static void performMigrations(final WeatherDatabase weatherDB, final LocationsDatabase locationDB) {
        Context context = SimpleLibrary.getInstance().getAppContext();
        long versionCode = 0;
        try {
            PackageInfo packageInfo = context.getPackageManager().getPackageInfo(context.getPackageName(), 0);
            versionCode = packageInfo.versionCode;
        } catch (Exception e) {
            Logger.writeLine(Log.DEBUG, e);
        }

        if (Settings.isWeatherLoaded() && Settings.getVersionCode() < versionCode) {
            // v4.2.0+ (Units)
            if (Settings.getVersionCode() < 294200000) {
                final String tempUnit = Settings.getTemperatureUnit();
                if (Units.CELSIUS.equals(tempUnit)) {
                    Settings.setDefaultUnits(Units.CELSIUS);
                } else {
                    Settings.setDefaultUnits(Units.FAHRENHEIT);
                }

                if (!SimpleLibrary.getInstance().getApp().isPhone()) {
                    Settings.setRefreshInterval(Settings.DEFAULTINTERVAL);
                }
            }

            if (Settings.getVersionCode() < 294310000) {
                if (WeatherAPI.HERE.equals(Settings.getAPI())) {
                    // Set default API to Yahoo
                    Settings.setAPI(WeatherAPI.YAHOO);
                    WeatherManager wm = WeatherManager.getInstance();
                    wm.updateAPI();

                    Settings.setPersonalKey(false);
                    Settings.setKeyVerified(true);
                }
            }

            if (Settings.getVersionCode() < 294320000) {
                // Update location keys
                // NWS key is different now
                if (Settings.IS_PHONE) {
                    DBUtils.updateLocationKey(locationDB);
                }
                Settings.saveLastGPSLocData(new LocationData());
            }

            if (Settings.getVersionCode() < 295000000) {
                if (WeatherAPI.YAHOO.equals(Settings.getAPI()) || WeatherAPI.HERE.equals(Settings.getAPI())) {
                    // Yahoo Weather API is no longer in service
                    // Set default API to WeatherUnlocked
                    Settings.setAPI(WeatherAPI.WEATHERUNLOCKED);
                    WeatherManager wm = WeatherManager.getInstance();
                    wm.updateAPI();

                    Settings.setPersonalKey(false);
                    Settings.setKeyVerified(true);
                }
            }

            Bundle bundle = new Bundle();
            bundle.putString("API", Settings.getAPI());
            bundle.putString("API_IsInternalKey", Boolean.toString(!Settings.usePersonalKey()));
            bundle.putLong("VersionCode", Settings.getVersionCode());
            bundle.putLong("CurrentVersionCode", versionCode);
            AnalyticsLogger.logEvent("App_Upgrading", bundle);
        }

        if (versionCode > 0) {
            if (Settings.getVersionCode() < versionCode) {
                FeatureSettings.setUpdateAvailable(false);
            }
            Settings.setVersionCode(versionCode);
        }
    }
}

File: shared_resources/src/main/java/com/thewizrd/shared_resources/utils/VersionMigrations.kt
package com.thewizrd.shared_resources.utils

import android.content.Context
import android.os.Bundle
import android.util.Log
import com.thewizrd.shared_resources.SimpleLibrary
import com.thewizrd.shared_resources.database.LocationsDatabase
import com.thewizrd.shared_resources.database.WeatherDatabase
import com.thewizrd.shared_resources.locationdata.LocationData
import com.thewizrd.shared_resources.preferences.FeatureSettings
import com.thewizrd.shared_resources.utils.DBUtils.updateLocationKey
import com.thewizrd.shared_resources.weatherdata.WeatherAPI
import com.thewizrd.shared_resources.weatherdata.WeatherManager

internal object VersionMigrations {
    suspend fun performMigrations(
        context: Context,
        weatherDB: WeatherDatabase,
        locationDB: LocationsDatabase
    ) {
        val settingsMgr = SettingsManager(context.applicationContext)

        val versionCode = try {
            val packageInfo =
                context.applicationContext.packageManager.getPackageInfo(context.packageName, 0)
            packageInfo.versionCode.toLong()
        } catch (e: Exception) {
            Logger.writeLine(Log.DEBUG, e)
            0
        }

        if (settingsMgr.isWeatherLoaded() && settingsMgr.getVersionCode() < versionCode) {
            // v4.2.0+ (Units)
            if (settingsMgr.getVersionCode() < 294200000) {
                val tempUnit = settingsMgr.getTemperatureUnit()
                if (Units.CELSIUS == tempUnit) {
                    settingsMgr.setDefaultUnits(Units.CELSIUS)
                } else {
                    settingsMgr.setDefaultUnits(Units.FAHRENHEIT)
                }
                if (!SimpleLibrary.instance.app.isPhone) {
                    settingsMgr.setRefreshInterval(SettingsManager.DEFAULTINTERVAL)
                }
            }

            if (settingsMgr.getVersionCode() < 294310000) {
                if (WeatherAPI.HERE == settingsMgr.getAPI()) {
                    // Set default API to Yahoo
                    settingsMgr.setAPI(WeatherAPI.YAHOO)
                    val wm = WeatherManager.instance
                    wm.updateAPI()
                    settingsMgr.setPersonalKey(false)
                    settingsMgr.setKeyVerified(true)
                }
            }

            if (settingsMgr.getVersionCode() < 294320000) {
                // Update location keys
                // NWS key is different now
                if (SimpleLibrary.instance.app.isPhone) {
                    updateLocationKey(context, locationDB)
                }
                settingsMgr.saveLastGPSLocData(LocationData())
            }

            if (settingsMgr.getVersionCode() < 295000000) {
                if (WeatherAPI.YAHOO == settingsMgr.getAPI() || WeatherAPI.HERE == settingsMgr.getAPI()) {
                    // Yahoo Weather API is no longer in service
                    // Set default API to WeatherUnlocked
                    settingsMgr.setAPI(WeatherAPI.WEATHERUNLOCKED)
                    WeatherManager.instance.updateAPI()
                    settingsMgr.setPersonalKey(false)
                    settingsMgr.setKeyVerified(true)
                }
            }

            val bundle = Bundle().apply {
                putString("API", settingsMgr.getAPI())
                putString("API_IsInternalKey", (!settingsMgr.usePersonalKey()).toString())
                putLong("VersionCode", settingsMgr.getVersionCode())
                putLong("CurrentVersionCode", versionCode)
            }
            AnalyticsLogger.logEvent("App_Upgrading", bundle)
        }

        if (versionCode > 0) {
            if (settingsMgr.getVersionCode() < versionCode) {
                FeatureSettings.setUpdateAvailable(false)
            }

            settingsMgr.setVersionCode(versionCode)
        }
    }
}